### PR TITLE
Do not ignore git push errors

### PR DESCRIPTION
**Pull request by sven-urbanski-freiheit-com**
Do not ignore git push errors
    
Also return readible error in case we cannot push to the manifest repo.
    
2 error cases are handled in this PR:
1) the ssh key does only have read only access. This returns a normal error (at least on github)
2) there is a branch protection rule stopping access. This is not a normal error, but signaled via the "PushUpdateReferenceCallback"

In both cases we return the error to the caller instead of "internal error".

In order to test these 2 error cases, I had to introduce 2 Callbacks for `ProcessQueueOnce`.

This is the solution for https://github.com/freiheit-com/kuberpult/issues/712

### CODE REVIEW

File: services/cd-service/Makefile [hunk 1]
```diff
@@ -77,7 +77,9 @@ run: docker
 	docker run --rm -e KUBERPULT_GIT_URL="/repository_remote" -e KUBERPULT_GIT_BRANCH=master -p 8080:8080 -p 8443:8443 -v $(shell pwd)/repository_remote:/repository_remote $(IMAGENAME)
 else
 run: build
-	KUBERPULT_GIT_URL="./repository_remote" KUBERPULT_GIT_BRANCH=master ./bin/main
+	#KUBERPULT_GIT_URL="./repository_remote" KUBERPULT_GIT_BRANCH=master ./bin/main
```
Comment (leandro-freiheit) [inline services/cd-service/Makefile:80]: Is these comment supposed to be here?

File: services/cd-service/pkg/repository/repository.go [hunk 2]
```diff
@@ -368,7 +393,9 @@ func (r *repository) ProcessQueueOnce(e element) {
 	}
 
 	// Try pushing once
-	err = r.Push(e.ctx, pushAction)
+	logger.Error(fmt.Sprintf("SU DEBUG before push"))
```
Comment (leandro-freiheit) [inline services/cd-service/pkg/repository/repository.go:396]: I think this should be removed

File: services/cd-service/pkg/repository/repository.go [hunk 3]
```diff
@@ -382,15 +409,20 @@ func (r *repository) ProcessQueueOnce(e element) {
...
+			err = httperrors.PublicError(ctx, errors.New(fmt.Sprintf("could not push to manifest repository '%s' on branch '%s' - this indicates that the ssh key does not have write access", r.config.URL, r.config.Branch)))
+		}
+	} else {
+		if !pushSuccess {
+			err = fmt.Errorf("failed to push - this indicates that branch protection is enabled in '%s' on branch '%s'", r.config.URL, r.config.Branch)
 		}
 	}
-	err = nil
 	r.notify.Notify()
+	logger.Error(fmt.Sprintf("SU DEBUG after notify"))
```
Comment (leandro-freiheit) [inline services/cd-service/pkg/repository/repository.go:425]: This should also be removed

File: services/cd-service/Makefile [hunk 4]
```diff
@@ -77,7 +77,9 @@ run: docker
...
+#	KUBERPULT_GIT_URL="git@github.com:sven-urbanski-freiheit-com/manifest-test.git" KUBERPULT_GIT_BRANCH="main" KUBERPULT_GIT_SSH_KEY="../../private.key" KUBERPULT_GIT_SSH_KNOWN_HOSTS=../../known_hosts_github  ./bin/main
+	KUBERPULT_GIT_URL="git@github.com:freiheit-com/fdc-standard-setup-dev-env-manifest.git" KUBERPULT_GIT_BRANCH="main" KUBERPULT_GIT_SSH_KEY="../../private.key" KUBERPULT_GIT_SSH_KNOWN_HOSTS=../../known_hosts_github  ./bin/main
```
Comment (hannesg) [inline services/cd-service/Makefile]: Is this change intended?

File: services/cd-service/pkg/repository/repository.go (hunk 3 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go:423]: done

File: services/cd-service/pkg/repository/repository.go [hunk 5]
```diff
@@ -328,7 +329,32 @@ func (r *repository) drainQueue() []element {
...
 
-func (r *repository) ProcessQueueOnce(e element) {
+// DefaultPushUpdate is public for testing reasons only.
+// It returns always nil
+// success is set to true if the push was successful
+func DefaultPushUpdate(branch string, success *bool) git.PushUpdateReferenceCallback {
+	return func(refName string, status string) error {
+		var expectedRefName = fmt.Sprintf("refs/heads/%s", branch)
+		// if we were successful the status is empty and the ref contains our branch:
+		*success = refName == expectedRefName && status == ""
```
Comment (hannesg) [inline services/cd-service/pkg/repository/repository.go:339]: Uh, this looks hard.

File: services/cd-service/Makefile (hunk 1 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/Makefile]: i removed them

File: services/cd-service/pkg/repository/repository.go (hunk 2 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go]: i removed it

File: services/cd-service/pkg/service/batch.go [hunk 6]
```diff
@@ -193,7 +192,7 @@ func (d *BatchServer) ProcessBatch(
 	}
 	err := d.Repository.Apply(ctx, transformers...)
 	if err != nil {
-		return nil, httperrors.InternalError(ctx, fmt.Errorf("could not apply transformer: %w", err))
+		return nil, err
```
Comment (hannesg) [inline services/cd-service/pkg/service/batch.go:195]: I'm a bit sad that this didn't trigger any tests.

**Review by leandro-freiheit — APPROVED** [review]

File: services/cd-service/pkg/repository/repository.go [hunk 7]
```diff
@@ -382,14 +407,18 @@ func (r *repository) ProcessQueueOnce(e element) {
...
 				return
 			}
-			if pushErr := r.Push(e.ctx, pushAction); pushErr != nil {
+			if pushErr := r.Push(e.ctx, pushAction(pushOptions, r)); pushErr != nil {
 				err = &InternalError{inner: pushErr}
 			}
 		} else {
-			err = &InternalError{inner: err}
+			logger.Error(fmt.Sprintf("error while pushing: %s", err))
+			err = httperrors.PublicError(ctx, errors.New(fmt.Sprintf("could not push to manifest repository '%s' on branch '%s' - this indicates that the ssh key does not have write access", r.config.URL, r.config.Branch)))
```
Comment (hannesg) [inline services/cd-service/pkg/repository/repository.go:415]: Most callers are not http but grpc. How will they display this error?

**Comment by hannesg** [comment]
I think you could construct a very realistic test-case by adding running the cd-service against a bare repo with an `update` hook.

File: services/cd-service/pkg/repository/repository.go (hunk 5 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go:339]: I don't know what this means, but I'm also not super happy with this code.
Do you think it won't work in some cases?
I tried it on GitHub and it works fine.

File: services/cd-service/pkg/repository/repository.go (hunk 7 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go:415]: the package name is misleading. It is actually returning the error fine for grpc, I tested it with `evans`.

**Comment by sven-urbanski-freiheit-com** [comment]
> I think you could construct a very realistic test-case by adding running the cd-service against a bare repo with an `update` hook.

I tested this manually with a real github repo.

File: services/cd-service/pkg/repository/repository.go (hunk 5 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go:339]: After lots of experimentation, I think I'll have to leave it as it is.
We tried running `git init` in a test with a server-side hook in the repo, to reproduce this case where the pushing is not successful. This would avoid this rather ugly callback with the `success` pointer. However, git hooks, incl server-hooks are not supported in this setup, and seem to generally not supported by gitlib2 https://github.com/libgit2/libgit2/blob/main/docs/projects.md?plain=1#L74
